### ISSUE TITLE
Census Custom Estimator Distributed Training not working with --num-epochs set

### ISSUE THREAD

**Issue by jcomfort4** [GoogleCloudPlatform/cloudml-samples#74]
### System information
- Mac OS submitting gcloud job
- 1.2.0
- 2.7

Command to reproduce: 
Running the CUSTOMESTIMATOR below
gcloud ml-engine jobs submit training $JOB_NAME \
                                     --scale-tier STANDARD_1 \
                                     --runtime-version 1.2 \
                                     --job-dir $GCS_JOB_DIR \
                                     --module-name trainer.task \
                                     --package-path trainer/ \
                                     --region us-central1 \
                                     -- \
                                     --train-files $TRAIN_FILE \
                                     --eval-files $EVAL_FILE \
                                     --num-epochs 1 \
                                     --eval-steps 100

### Describe the problem
If I try to run the distributed mode command above with --num epochs set to 1 then I run into an error shortly after training begins. However, if I don't set num-epochs and I just set --train-steps (As in the sample) the model appears to train just fine. Is anybody able to reproduce the behavior? I realize it easily could be my own human error. However, when I ran it with both --train-steps 1000 and --num-epochs 1 I was able to get the code to succesfully complete (though some times with the same settings it would not complete). This makes me think that there is something in the input data causing this error (labels must be 1D) which would explain the random chance of completing or not completing.

## Traceback

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/local/google/home/elibixby/repos/GoogleCloudPlatform/cloudml-samples/census/customestimator/trainer/task.py", line 209, in <module>
    hparams=hparam.HParams(**args.__dict__)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/learn_runner.py", line 210, in run
    return _execute_schedule(experiment, schedule)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/learn_runner.py", line 47, in _execute_schedule
    return task()
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 495, in train_and_evaluate
    self.train(delay_secs=0)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 275, in train
    hooks=self._train_monitors + extra_hooks)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 660, in _call_train
    hooks=hooks)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 241, in train
    loss = self._train_model(input_fn=input_fn, hooks=hooks)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 612, in _train_model
    _, loss = mon_sess.run([estimator_spec.train_op, estimator_spec.loss])
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/training/monitored_session.py", line 505, in run
    run_metadata=run_metadata)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/training/monitored_session.py", line 842, in run
    run_metadata=run_metadata)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/training/monitored_session.py", line 798, in run
    return self._sess.run(*args, **kwargs)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/training/monitored_session.py", line 952, in run
    run_metadata=run_metadata)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/training/monitored_session.py", line 798, in run
    return self._sess.run(*args, **kwargs)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 789, in run
    run_metadata_ptr)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 997, in _run
    feed_dict_string, options, run_metadata)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1132, in _do_run
    target_list, options, run_metadata)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/client/session.py", line 1152, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: labels must be 1-D, but got shape []
	 [[Node: SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits = SparseSoftmaxCrossEntropyWithLogits[T=DT_FLOAT, Tlabels=DT_INT64, _device="/job:localhost/replica:0/task:0/cpu:0"](dense_5/BiasAdd, Squeeze)]]

Caused by op u'SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits', defined at:
  File "/usr/lib/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/usr/local/google/home/elibixby/repos/GoogleCloudPlatform/cloudml-samples/census/customestimator/trainer/task.py", line 209, in <module>
    hparams=hparam.HParams(**args.__dict__)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/learn_runner.py", line 210, in run
    return _execute_schedule(experiment, schedule)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/learn_runner.py", line 47, in _execute_schedule
    return task()
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 495, in train_and_evaluate
    self.train(delay_secs=0)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 275, in train
    hooks=self._train_monitors + extra_hooks)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/contrib/learn/python/learn/experiment.py", line 660, in _call_train
    hooks=hooks)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 241, in train
    loss = self._train_model(input_fn=input_fn, hooks=hooks)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 560, in _train_model
    model_fn_lib.ModeKeys.TRAIN)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 545, in _call_model_fn
    features=features, labels=labels, **kwargs)
  File "trainer/model.py", line 196, in _model_fn
    logits=logits, labels=label_indices_vector))
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/ops/nn_ops.py", line 1690, in sparse_softmax_cross_entropy_with_logits
    precise_logits, labels, name=name)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/ops/gen_nn_ops.py", line 2486, in _sparse_softmax_cross_entropy_with_logits
    features=features, labels=labels, name=name)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/framework/op_def_library.py", line 767, in apply_op
    op_def=op_def)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 2506, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/usr/local/google/home/elibixby/envs/tf1.2/local/lib/python2.7/site-packages/tensorflow/python/framework/ops.py", line 1269, in __init__
    self._traceback = _extract_stack()

InvalidArgumentError (see above for traceback): labels must be 1-D, but got shape []
	 [[Node: SparseSoftmaxCrossEntropyWithLogits/SparseSoftmaxCrossEntropyWithLogits = SparseSoftmaxCrossEntropyWithLogits[T=DT_FLOAT, Tlabels=DT_INT64, _device="/job:localhost/replica:0/task:0/cpu:0"](dense_5/BiasAdd, Squeeze)]]
```

**Comment by jcomfort4** [comment]
For reference, I am trying to see if I can replicate the behavior observed [here](https://stackoverflow.com/questions/45989971/distributed-training-with-tf-estimator-resulting-in-more-training-steps) on the custom estimator code, but obviously not being able to run for a full epoch is not allowing me to repeat this analysis on different distributed training configurations.

**Comment by elibixby** [comment]
For future reference, single line tracebacks are impossible to read. Please preserve formatting from wherever you pasted that ;-). I've replicated your issue, and replaced your traceback with mine.

**Comment by jcomfort4** [comment]
Sorry! Noted for future reference. Thank you for reproducing and improving my traceback :)

**Comment by elibixby** [comment]
Fixed in #75 See the PR for my diagnosis =)

**Comment by puneith** [comment]
@jcomfort4 This has been merged with master now. Thanks @elibixby.
